### PR TITLE
migrate to anywidget

### PR DESCRIPTION
**Pull request by manzt**
Flight got delayed so I'm stuck in the airport and had some time on my hands lol.

This PR migrates `jupyter-scatter` to be an **anywidget** Jupyter Widget. I think this was a Goal we'd discussed a little while ago, so I just went ahead with it. Would love to have you try it out and see what you think.

The JS Code took some refactoring, but you can see the diff in `widget.py` is super minimal. The biggest change now is that there is no longer a JS package to publish to NPM. All the JS source is included in the code that you publish to PyPI.

Changes:

- Single JS entrypoint (`js/src/index.js`) that exports `render`, which is expected by anywidget.
- No more webpack, just use `esbuild` to bundle `js/src/index.js` -> `jscatter/bundle.js`). Jupyter-scatter has several npm dependencies, and while although we could just import them from a CDN and forgoe any bundling, this setup ensures jscatter will work offline too. 
- Switch from setuptools to hatching build system. Not essential, but it's the recommend build system now and just has nicer defaults (e.g., no need for `MANIFEST.in`). Also, I configured it to just read the version directly from `package.json` for the python package.
- Consolidate all python metadata into `pyproject.toml` (removes `requirements.txt`, `requirements-dev.txt`, `setup.py`, and `setup.cfg`). Kind of a stylistic choice, but it's nice to have all this in one place.

Things that need further investigation/discussion:

- What to do with `jscatter/bundle.js`? I would just recommend checking this file into the repo. Otherwise you'll still need to setup your python build system to build the js. It really isn't necessary and checking in the file is way simplier imo.
- The `JsScatterView.destory` is throwing some errors in the browser that I can't seem to figure out. It seems to be deleting the canvas in `regl-scatter-plot`?

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -28,3 +28,4 @@ jscatter/labextension
 jscatter/nbextension
 
 notebooks/data
+jscatter/bundle.js
```
Comment (manzt) [inline .gitignore:31]: Ignoring for now, but probably should check in the production build.

**Comment by manzt** [comment]
The size of the diff here is dominated by `package-lock.json`

**Comment by manzt** [comment]
> Does this also mean I can safely and fully embrace ES6 now?

Yup exactly. `esbuild` [targets `esnext`](https://esbuild.github.io/api/#target) by default, so we can use whatever we want in src/index.js and it won't be transpiled. We could target something else if we want but it should be fine.

> I'm happy to follow your recommendation here. But just to clarify, what is this bundle.js file? Is it equivalent to the current webpack output? Or is this a new bundle file? Also, would this have to change every time we release a new version?

`bundle.js` is the ESM bundler output of the contents of `js/src`. It's created with esbuild via: `esbuild --bundle --formet=esm --outfile=../jscatter/bundle.js src/index.js`. anywidget requires a true ESM entrypoint, so we need to have a step that injects the imports from `node_modules` so that there is a file to give to anywidget that it can load in the browser. An alternative would be to have imports come from a CDN, e.g.,

```js
import * as d3Scale from "https://esm.sh/d3-scale@3";
///
```

and then anywidget could load `src/index.js` directly. But, then the frontend code wouldn't work offline, and that's something you've told me is important to you and this project, so having a build step is necessary. It's just now there is only one target (`bundle.js`), which is ESM and uses latest js syntax.

> Mhhh this is kind of critical to get right. The destroy function should clear all resources related to the view instance but it should not destroy the global renderer.

Yup, I can't look into it for a couple of days but will have a closer look!

**Comment by flekschas** [comment]
> But, then the frontend code wouldn't work offline, and that's something you've told me is important to you and this project, so having a build step is necessary.

That makes sense. And yeah, while most people are likely online when they use jscatter, I think the plugin should work offline too.

The one thing I still like to understand better is why/if we have to check the bundled file into the repo instead of just uploading it to npm. This is how it works currently: [the bundled file only lives on npm](https://cdn.jsdelivr.net/npm/jupyter-scatter@0.11.0/). I find that to be a very nice separation of concerns. Do you know if this is possible with anywidget?

**Comment by manzt** [comment]
> This is how it works currently: the bundled file only lives on npm. 

This isn't true. There are multiple bundles that are generated by the webpack config to target Jupyterlab, jupyter notebooks and npm for use of the widget via a cdn (colab/vscode). Only the latter is published to NPM. The other bundles are placed in your file system for jupyter/jupyterlab to discover as extension.

Discovery of extensions is a huge headache and requires complicated packaging to get right. With anywidget, you point your python class directly to the code it needs to load from the file system, and all environments (vscode/colab/jupyter) reuse the same module.

**Comment by flekschas** [comment]
Yeah let's jump on a call when you're back. :) I might be missing something. But if possible I'd like to avoid checking bundled assets into the repo.

**Comment by manzt** [comment]
Ok! I resolved the merge conflicts here and added a build step to the `pyproject.toml` to generate `jscatter/bundle.js`.

File: pyproject.toml [hunk 2]
```diff
@@ -1,3 +1,73 @@
...
+[tool.hatch.build.hooks.jupyter-builder]
+build-function = "hatch_jupyter_builder.npm_builder"
+ensured-targets = ["jscatter/bundle.js"]
+skip-if-exists = ["jscatter/bundle.js"]
+dependencies = ["hatch-jupyter-builder>=0.5.0"]
+
+[tool.hatch.build.hooks.jupyter-builder.build-kwargs]
+npm = "npm"
+build_cmd = "build"
+path = "js"
```
Comment (manzt) [inline pyproject.toml:73]: This is a plugin for hatchling that runs the `npm_builder` when you run `pip install -e .` or `python -m build`. If `jscatter/bundle.js` exists (what we publish to pypi), then the npm build step is skipped.

**Comment by flekschas** [comment]
Cool, thanks! I'll give it a try this afternoon

File: js/src/codecs.js [hunk 3]
```diff
@@ -17,7 +17,7 @@ const DTYPES = {
  * @prop {Shape} shape
  */
 
-export default function Numpy2D(dtype) {
```
Comment (manzt) [inline js/src/codecs.js]: @flekschas oops, this was the issue...! should be working now

**Review by flekschas — APPROVED** [review]
Everything seems to be working now 🥳

**Comment by manzt** [comment]
huzzah :) Wish that was a bit less painful but soo happy that we were able to make the migration!

**Pull request merged**